tests/upgrades/test_containers_removal.py

```diff
@@ -92,60 +92,64 @@ def test_pre_scenario_containers_support_removal(self):
-        self.assertEqual(running_containers.return_code, 0)
-
-        self.assertTrue(any(dockerhub_container.name in line
-                            for line in running_containers.stdout))
-        self.assertTrue(any(external_container.name in line
-                            for line in running_containers.stdout))
+            # Only one registry with given URL can exist on Satellite,
+            # so search for it first and create it only if necessary
+            try:
+                registry = entities.Registry().search(filters={'url': registry_url})[0]
+            except IndexError:
+                registry = entities.Registry(
+                    url=registry_url,
+                    organization=[org],
+                ).create()
+
+            # container from dockerhub
+            dockerhub_container = entities.DockerHubContainer(
+                command='sleep inf',
+                compute_resource=compute_resource,
+                organization=[org],
+            ).create()
+            self.assertEqual(
+                dockerhub_container.compute_resource.id, compute_resource.id)
 
-        scenario_dict = {self.__class__.__name__: {
-            'docker_host': docker_host.hostname,
-            'dockerhub_container': dockerhub_container.name,
-            'external_container': external_container.name,
-        }}
-        create_dict(scenario_dict)
+            # container from external registry
+            external_container = entities.DockerRegistryContainer(
+                command='sleep inf',
+                compute_resource=compute_resource,
+                organization=[org],
+                registry=registry,
+                repository_name=repo_name,
+            ).create()
+            self.assertEqual(
+                external_container.compute_resource.id, compute_resource.id)
+            self.assertEqual(external_container.registry.id, registry.id)
+            self.assertEqual(external_container.repository_name, repo_name)
+
+            running_containers = docker_host.run('docker ps')
+            self.assertEqual(running_containers.return_code, 0)
+
+            self.assertTrue(any(dockerhub_container.name in line
+                                for line in running_containers.stdout))
+            self.assertTrue(any(external_container.name in line
+                                for line in running_containers.stdout))
+
+            scenario_dict = {self.__class__.__name__: {
+                'docker_host': docker_host.hostname,
+                'dockerhub_container': dockerhub_container.name,
+                'external_container': external_container.name,
+            }}
+            create_dict(scenario_dict)
+        except Exception as exp:
+            self._vm_cleanup(hostname=docker_host.hostname)
+            raise Exception(exp)
 
     @post_upgrade(depend_on=test_pre_scenario_containers_support_removal)
     def test_post_scenario_containers_support_removal(self):
```

Comment:
@mirzal Do you think same try catch also should be applied to the postupgrade method as well? I can see the earlier example you gave from Repo Don't have.

Reply:
Exactly, test_remoteeecution doesn't have it.

I believe the idea is that if post_upgrade failed, it's probably due to error in environment or test, so we fix it quickly and re-run this part. This way we don't have to run pre-upgrade and costly upgrade process for every issue we encounter. So we want to keep VM around until we explicitly remove it.

Reply:
Hmm Yeah, that makes sense.

Reply:
Fair point. Added in b0e6c4b50b0f53bf7506a1227122ecd4a942a3f7